File: regularexpressionmatching.java
public class Solution {
    
    
    
    public boolean isMatch(String s, String p) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if(s.length()==0 && p.length()==0 ) {
            return true;
        }
        else if(p.startsWith(".*")){
            for(int i=0;i<=s.length();i++){
                if(isMatch(s.substring(i),p.substring(2))){
                    return true;
                }
            }
            return false;
        }else if(p.startsWith(".")){
            if(s.length()==0) return false;
            return isMatch(s.substring(1),p.substring(1));
        }else if(p.length()>=2 && p.charAt(1)=='*' ){
            
            if(s.length()==0){
                return isMatch(s,p.substring(2));
            } 
            
            int i=0;
            
            if(isMatch(s.substring(0),p.substring(2))){
                return true;
            }
            
            while(i<s.length() && p.charAt(0)==s.charAt(i)){
                if(isMatch(s.substring(i+1),p.substring(2))){
                    return true;
                }
                i++;
            }
            
            return false;
        }else if(p.length()>=1 ){
            if(s.length()==0) return false;
            
            if(s.charAt(0)!=p.charAt(0)){
                return false;
            }
            return isMatch(s.substring(1),p.substring(1));
        }else{
            return false;
        }
    }
}

File: cpp/RegularExpressionMatching.cpp
class Solution {
public:
    bool isMatch(const char *s, const char *p) {
        if(s==NULL || p==NULL) {
            return false;
        }
        if(strlen(p)==0 && strlen(s)==0) {
            return true;
        }
        if(strlen(p)==0 && strlen(s)>0) {
            return false;
        }
        if(strlen(p)==1) {
            if(p[0]=='*') {
                return false;
            }
            else if(strlen(s)==0) {
                return false;
            }
            else if(p[0]=='.') {
                return isMatch(s+1, p+1);
            }
            else {
                return p[0] == s[0] && isMatch(s+1,p+1);
            }
        }
        else {
            if(p[0] == '.' && p[1] == '*') {
                while(true) {
                    if(isMatch(s, p+2)) {
                        return true;
                    }
                    if(*s=='\0') {
                        return isMatch(s,p+2);
                    }
                    s++;
                }
                return false;
            } else if(p[1] == '*') {
                char t = p[0];
                if(isMatch(s,p+2)) {
                    return true;
                }
                while(*s == t) {
                    if(isMatch(s+1,p+2)) {
                        return true;
                    }
                    s++;
                }
                return false;
            } else if(p[0] == '.') {
                if(strlen(s)==0) {
                    return false;
                }
                return isMatch(s+1, p+1);
            } else {
                if(strlen(s)==0) {
                    return false;
                }
                return (s[0] == p[0]) && isMatch(s+1,p+1);
            }
        }
    }
};
